### DISCUSSION TITLE
Parallel to SPI conversion?

### DISCUSSION THREAD

**Discussion by maleksaid** [bjaan/roland-juno-g-display-replacement#6] [Q&A]
Very interesting project. I am trying to do the same thing but with the Roland Fantom XR where I have a dead LCD. Do bear with me as this is all (dealing with LCDs) pretty new to me.

Correct me if I am wrong, it appears that you have converted the Roland Juno G's parallel output (to LCD) to an SPI output to the new LCD unit.

Is the parallel data (D0 to D7) from the Juno G standard? Surely we need to know how the image is structured in order to convert it to SPI. Is there a standard algorithm to convert it from parallel data to SPI or we need to create new algorithms for different keyboards?

**Comment by bjaan** [comment]
Hi @maleksaid,

In this project, the Pico software converts the signals coming from the Roland Juno G into SPI signals that the (color) LCD screen understands.

No, it is not just parallel data (D0 to D7), since it also includes two extra connections (CS1 and CS2). The original Juno G display has dual Samsung KS0713 display controllers, each controlling one of the two parts of the screen (i.e., the left and right parts). The Pico is now emulating this setup, with two additional CS (Chip Select) inputs to control each part, managed by an emulated KS0713. This allows it to update either part of the two-part screen.

When you want to make this work, you first need to determine what type of display the Fantom XR uses, or more specifically, what type of display controller is used in the Fantom XR.

This is what I see in its service notes: it uses the following display module, which integrates both the display and the display controller.  
![{BE0BFE42-7A07-4634-8B1F-9D67383058E1}](https://github.com/user-attachments/assets/1cfecd15-a4fa-4e14-9ccb-1cf14a0478db)

Maybe the display is this one from [http://lcddisplaymodules.com](http://lcddisplaymodules.com/e_productshow/?156-160x48-LCD-Display-156.html).

The display (on the right) is connected via a flat cable to the Fantom XR's main board (on the left), with a sub-board in between (middle):  
![{FAEB029D-6AA0-45CA-8547-243740207283}](https://github.com/user-attachments/assets/f8569d63-e75a-498d-8f9a-6f0a6339c462)

If the link above is correct, the display controller is the UltraChip UC1604C. You can find the datasheet here: [https://admin.osptek.com/uploads/uc1604c_a1_3_efad46687f.pdf](https://admin.osptek.com/uploads/uc1604c_a1_3_efad46687f.pdf).

These would serve as inputs for the Pico:  
![{9C185C71-2442-41B8-9C81-434F1CD5DED4}](https://github.com/user-attachments/assets/2347aba4-a5cf-4c4b-a446-039ab8a6ca73)

However, there is more in the service notes for the Fantom XR. This is how the side of the main board looks — if I understand correctly. It seems like M1P11 connections are being used, but I am not entirely sure; it could also be M1P10.  
![afbeelding](https://github.com/user-attachments/assets/3798971a-0ea2-44f3-ba8d-a22b0b26fbec)

This is the "panel" board:  
**Left side of the panel board:**  
![{8E2C7EE3-2892-4BDC-BDB6-9C3E0F23500D}](https://github.com/user-attachments/assets/55fac182-ac22-4e40-b63d-287bbe3ad017)

**Right side of the panel board:**  
![{E52162CE-79DB-4517-97C5-BE29EB364C01}](https://github.com/user-attachments/assets/192d2f98-e82b-49e3-b2ae-931004ff2a17)

The right side of the panel board seems to be the one to emulate, as I believe this connects to the display module. One flat cable carries data and voltages, while another connects to the fixed backlight.

I am not sure if WR0 and WR1 are connected to Vss and whether the display is running in serial mode or parallel mode. This warrants further research.

Check this response: [https://github.com/bjaan/roland-juno-g-display-replacement/issues/1#issuecomment-1356801012](https://github.com/bjaan/roland-juno-g-display-replacement/issues/1#issuecomment-1356801012). It is an attempt to explain how it works for the Juno G.

**Reply by maleksaid** [reply]
Thanks bjaan for your response and the effort that you have put in on this matter.

For a start, let's discuss on the LCD. I have attached below some pictures of the original (broken) LCD.
![WhatsApp Image 2024-10-18 at 11 30 11(2)](https://github.com/user-attachments/assets/75b64844-a49c-4047-a1d5-719da0a08de5)
![WhatsApp Image 2024-10-18 at 11 30 11(1)](https://github.com/user-attachments/assets/cddb3b24-0d52-4ee9-8f71-7587be7f5bb8)

Given that Fantom XR uses an LCD with an **_unusual_** resolution of 160x48, there is a possibility that the link that you gave above (or is it below?) might be the a replacement LCD. However, the looking at the number of pins (14 vs 24) and the different connector for backlight **may** suggest that they may be different.

Now, if you look at the schematics, **LCDC** (on the mainboard) is an M66273FP. Checking the datasheet, M66273FP is an LCD controller with VRAM. Is it possible to have an LCD controller on the mainboard and another lcd controller on the LCD?

Looking at LCDC on the schematics, it appears that only VD0 to VD4 is used. Does this mean that it is only a 4-bit parallel data? Other connections that goes to the sub board are FLM, LP, CP and LCDENB. These (together with VD0 to VD4) becomes BBD0 to BBD7 on the sub board and then becomes D0 to D7 on the LCD connector?


With regards to the connection from the mainboard to the sub board, it uses the M1P10. Looking at the mainboard, it appears that the connection points for M1P10, M1P03 and M1P11 is at the same location. It is the socket that is soldered to the board that determines the type of connection that is being used (i.e. if CN8 is used, then it is M1P10. If CN20 and CN21 are soldered to the mainboard, then it uses M1P11 etc). For my case, CN8 with 28 PINs is soldered to the mainboard, hence M1P10. I have attached some pics below
![WhatsApp Image 2024-10-18 at 11 30 10(1)](https://github.com/user-attachments/assets/3502c817-555b-450b-a7b2-28467b21dcc9)
![WhatsApp Image 2024-10-18 at 11 30 10](https://github.com/user-attachments/assets/9f126ed0-3962-43c2-9b37-d6ea1fc044d5)


Are my observations above correct? Any help, pointers are much appreciated.

**Comment by bjaan** [comment]
In your last photo, I can see that the LCD module's data and voltages flat cable are connected to CN6 and the power for the background light is set to CN11, so that matches with the picture marked with **Right side of the panel board** in my earlier response.

Next to CN6 is also indicated _24FLT-SM1-TB_
![image](https://github.com/user-attachments/assets/4e74b125-7cc6-4214-92cd-9571e6e0cad5)

I think you can power the Pico using 5V coming from the CN6 connector
![{6F570C51-3AA8-4B1F-8C77-F9739789E9D7}](https://github.com/user-attachments/assets/c80e0aa3-f04c-40ba-a5e2-9677c06be959)

Regarding the schematics, it looks like the M66273FP is indeed marked as LCDC, probably LCD Controller, and I think it might act analogous to a GPU in a PC (i.e. character generation and other graphic functions), where the current unknown controller chip behind the LCD acts as the controller of the individual pixels on the LCD panel.

I find this in the documentation of the M66273FP, it has a LCD Interface, which is probably matching with the connections going into the LCD module.

![{41506CFF-8D04-4C54-823A-BE87842505F9}](https://github.com/user-attachments/assets/8920a0a9-234e-4817-ba5f-036961ff935e)

This LCDC is connected with this bus like this:
![image](https://github.com/user-attachments/assets/92121708-d2e0-4fa3-bfc2-1d338367bd66)

And when we go back the M66273FP LCD Interface table, data lines VD0 through VD7 can be connected in either 4-bit or 8-bit parallel, and as VD4-VD7 are marked as not connected, I see that Roland went with 4-bit parallel.

Yes, the **CN8** in this diagram proves that it is indeed the M1P10 - and this BAN CARD refers to this the actual flat cable, it seems, as it unclear I only found one result on google and specs in Japanese:
![image](https://github.com/user-attachments/assets/d6c92d00-a5bc-4b3a-9b87-5dfb9231fcdd)

However, D0 through D7 refer to the specific interface of M66273FP LCD Interface table like this, from left to right:
![image](https://github.com/user-attachments/assets/f335c72b-132d-4468-b6de-d26ecada93dc)

Or in a clear table:
| CN6 Pin Number | CN6 Pin Marking | LCD Input Pin  | LCD Input Pin Description  |
| :---:   | :---: | :---: | :---:  |
| 7 | D0   | LCDENB   | **LCD (ON/OFF) control signal output**<br>Output data which is set at bit "0" of mode register (R1) in the control register. This signal can be used for controlling the LCD power supply, because LCDENB is set to "L" by RESET
| 8 | D1   | CP  | **Display data transfer clock**<br>Shift clock for the transfer of display data to LCD.<br>Take the display data of VD<n:0> to LCD at falling edge of CP
| 9 | D2   | LP   | **Display data latch pulse**<br>This clock use both as the latch pulse of display data for LCD and the transfer of scanning signal.<br>LP is output when it finishes transferring display data of a line.<br>Latch of display data and the transfer of scanning signal at falling edge of LP
| 10 | D3   | FLM   | **First Line Marker signal output**<br>Output the start pulse of scanning line.<br>This signal is "H" active, the IC for driving scanning line catches FLM at falling edge of LP.
| 11 | D4   | VD0   | **Display data bus for LCD**<br>Transfer the LCD display data in synchronization with a rising edge of CP by putting 4-bit (or 8-bit) in parallel.<br>The VD<n:0> output pin in use differs depending on the number of driven screens and the display mode.<br>DATA BIT 0
| 12 | D5   | VD1   | DATA BIT 1
| 13 | D6   | VD2   | DATA BIT 2
| 14 | D7   | VD3   | DATA BIT 3

Can you maybe find out what type of controller chip is behind the LCD Panel? It looks like it is easy to disassemble the LCD module, using these plastic tabs.

**Reply by maleksaid** [reply]
Thanks once again bjaan for your incredible work.

Yes we are on the same page as to the connection from the MS66273FP right up to CN6 (on the sub board) which connects to the LCD module.

As you have suggested, I have removed the LCD module. Here are the pics (front and back)
![WhatsApp Image 2024-10-18 at 21 57 35](https://github.com/user-attachments/assets/0d57d2a8-d761-4220-8e7f-8a5df280e4d2)
![WhatsApp Image 2024-10-18 at 21 56 32](https://github.com/user-attachments/assets/737d0bd1-251e-4e3a-a277-4dcebd004438)

The controller chip is probably behind that black rubbery blob (in the 2nd pic). I'll try to scrape it off tomorrow with better lighting (it is 10:30pm here in Malaysia) and see if I can identify the controller chip.


I know I may be jumping the gun a bit but for my knowledge & education, one of the following 3 scenarios may happen:

1. We can identify the controller chip but we **cannot** find a LCD module with a resolution of 160x48. Can we use an LCD module with same controller chip but with a higher resolution (256x64 perhaps?)
2. We **cannot** identify the controller chip. Based on the PINOUT on CN6, can PICO convert the signal for the UC1604C (or any other chip) which has the same 160x48 LCD resolution?
3. Finally, same as (2) above but using an LCD with a higher resolution (256x64)?
 

Will revert tomorrow if I am successful (or otherwise) in identifying the controller chip on the LCD.

**Comment by bjaan** [comment]
I am not sure, if removing the black rubber will reveal a single chip too. I could be a bunch of them. Maybe, you should try to apply some heat to soften the rubber/resin.

What this project is doing is up-scaling (enhance 2 times) the original display from the Juno-G, as the connected new display has a higher native resolution, so I think you just need adjust this in the code, of course you can start with 1 pixel is 1 pixel and the outside area will appear blank.

Also, I found a similar LCD module with that interface.  However it is not 24-pin, and the important pins match

![{A18289BB-BC1E-4700-83E9-FE3A1FEA7C5C}](https://github.com/user-attachments/assets/81dd3f2f-3855-4101-9310-781742433026)
![{CAEA81B9-1FA2-49C6-B1EF-84264B4D8CDC}](https://github.com/user-attachments/assets/c035144f-88f7-42b8-85c3-265327f00e08)

Datasheet https://datasheet.octopart.com/DMF-682AN-EW-Optrex-datasheet-24516.pdf

Via https://electronics.stackexchange.com/questions/582811/lcd-does-not-have-register-select-or-read-write-pins-instead-has-flm-lp-cp-a

**Reply by maleksaid** [reply]
Managed to gently scrape off the rubber blob. I just applied some heat (about 130C) with a heatgun and the rubber came off easily.

Having cleaned most of the rubber, I found only one piece of metal (probably some kind of chip) with no markings. See the attached pics (front and back)
![WhatsApp Image 2024-10-19 at 15 34 13](https://github.com/user-attachments/assets/72384f37-6fab-4ad8-b0b4-b1422b483e72)
![WhatsApp Image 2024-10-19 at 15 34 42](https://github.com/user-attachments/assets/7f7be0f3-ceb7-461e-bee3-266b8801ebb8)

Thanks for finding the DMF682AN-EW Optrex LCD Module. At USD100 each, it costs almost the same as a second hand Fantom XR. However, I did find an alternative [here](https://www.sccdisplay.com/product/lcd-module-display-jxd320240a/). May give that a shot.

What are your thoughts having seen the pics above?

Do you reckon it is still possible for me to do a project to replace the LCD with and OLED?

**Comment by bjaan** [comment]
Yes, I am still pretty sure that it (i.e. converting the original display's signals to a new display that uses different signals) can be done in using the Pico as long as the PIO State Machine is fast enough to handle them, and as you can connect any type of screen (OLED and LCD) that is compatible with a Pico, and as long as it has a large enough resolution.  However, the code doing the conversion for the Juno-G's display needs to be adjusted for the Fantom XR's diplay as it using different signals.

Unfortunately, peeling away the rubber did not reveal any numbers, so this still is an unknown chip.  You either put it under a microscope to further reverse engineer, or capture the signals Fantom XR using a logic analyzer and try to write a Python script, like original author of this method _dpeddi_ used, or similar to try and emulate the screen, or take some ideas from similar datasheets.

The DMF682AN-EW Optrex LCD Module documentation can help with writing the new code, as it is controlled in the same manner as the Fantom XR's one.
![{EC22A739-7788-4A9E-85B2-AF05FC61253D}](https://github.com/user-attachments/assets/187bb622-4481-4a61-8284-47a8d20968e4)

It looks like that PIO microcode should be used as well: currently there are two PIO programs _LCDJunoG_cs1_program_ and _LCDJunoG_cs2_program_, where it looks like there is only one necessary for the Fantom XR's display: e.g. _LCDFantomXR_lp_program_ as it needs to wait for the LP pin to go from High to Low, and then push the D1, D2, D3, and D4 data bit pins into the ISR (Internal Shift Register) to store a full one frame of the display, which you then read using C++ code, this is what I think - untested:

**PIO Setup**
```c
  lcdFantomXR_lp.begin(2, pio0, 1);
  lcdFantomXR_lp.read_async(buffer);
```
**Modified display code**
```c
  if(latest_packet_timestamp == rolandFantomXR_lp.latest_packet_timestamp()) {
    return; // no packet received
  } 
  latest_packet_timestamp = rolandFantomXR_lp.latest_packet_timestamp();
  tft.startWrite();
  uint i = 0;
  for (uint y = 0; y < 48; y++)
  { 
    for (uint x = 0; x < 160; x+=4)  
    { 
      byte val = buffer[i];
      if (back_buffer[i] != val)
      {
       back_buffer[i] = val;
       drawPixel(val && 0001, x, y);
       drawPixel(val && 0010, x+1, y);
       drawPixel(val && 0100, x+2, y);
       drawPixel(val && 1000, x+3, y);
     }
     i++;
  }    
  tft.endWrite();
```

**Reply by maleksaid** [reply]
Thanks bjaan.
Do bear with me as I am now studying the programing codes and trying to understand it as C++ is new to me. The last time I did any programming was in University where we used Fortran 77.

Juno G's LCD has two KS0713 controllers. Hence the need to have cs1 and cs2 (and pio0 and pio1?)
So if the LCD has only one display controller, we only need cs1. Is my understanding correct?

With reference to the following code:
**_lcdFantomXR_lp.begin(2, pio0, 1)_**
This basically begins the capture of D0 to D3 (and subsequently to push it to ISR) when LP goes from High to Low?
What if we want to capture D0 to D3 when LP goes from Low to High? Where do I need to amend the code?

**Reply by bjaan** [reply]
Yes, in the case of Fantom XR, there is just one signal, LP, to monitor instead of two, CS1 and CS2, so there going to be just one PIO program (_.program_ bookmark), see the _LCDJunoG.pio_ file in this repository.

```pio
.program LCDJunoG_cs1

.wrap_target
    wait 0 pin 10             ; cs1
    wait 1 pin 10                      ; Stall until start bit is asserted

bitloop:
    in pins, 10                        ; Shift data bit into ISR
        
    in NULL, 22                       ; Push 24 more bits into the ISR so that our one byte is at the position where the DMA expects it
    push                              ; Should probably do error checking on the stop bits some time in the future....

.wrap
```

You will need to read 4 pins instead of one (D0, D1, D2, and D3) for a number of times.  That is what the _in_ instruction does, you can have more than 1 pin there, and that should be pins in the case of the pins connected to D0-D3

Have a look at this lecture on PIO to understand what is going on https://www.youtube.com/watch?v=wet9CYpKZOQ  I am not an expert on PIO code programming, and used this code from the original author.

**Reply by maleksaid** [reply]
I have been studying the main.cpp code and I am beginning to understand what it is trying to do. However, I am a bit confused when it comes to the GPIO pins.

For example, the following code in LCDJunoG.pio
<img width="424" alt="wrap" src="https://github.com/user-attachments/assets/d3e97c4d-bf6e-4bdd-9392-1cd650332245" />

Based on the above cs1 is pin 10 or GP10 (and cs2 is pin 11 or GP11). However, when I cross checked to the PINOUT diagram below, CS1 is on GP13, which is different to the above

<img width="293" alt="pins" src="https://github.com/user-attachments/assets/5a08b62c-4a5b-4135-9dd4-6ad027fb9be1" />

Similarly, the above diagram shows that RS is on GP11. However, the following code in main.cpp says it is on GP6
<img width="392" alt="buffer" src="https://github.com/user-attachments/assets/aaa7a2bc-e7c1-4868-98a8-bb9bab9d07b7" />

Am I looking at it correctly? Appreciate if you could shed some light as I am thoroughly confused right now.

Also, based on the code above, buffer_cs1[i] may be represented as ??????s? vvvvvvvv
Am I correct to deduce that the 16 bit can be mapped to GPIO pins as follows:
<img width="455" alt="table" src="https://github.com/user-attachments/assets/cbe8bed9-1e8f-4028-a437-aa0e25438a12" />


Did I get the arrangement right? Appreciate any pointers you can give.

**Comment by bjaan** [comment]
The LCD connector looks like a 24pin 0.5 mm pitch FPC.
Using this cable
https://www.amazon.com/FOCMKEAS-Flexible-Ribbon-Printer-Type/dp/B0C5TL9SSF
And, this board you can break out the signals to DuPont connectors, so you can easily couple the Fantom XR to the Pico or Logic Analyzer:
https://www.amazon.com/FPC-24PIN-2-54MM-Adapter-Connector-Converter/dp/B09T5JTYQV

**Reply by maleksaid** [reply]
I actually bought the 24pin 0.5mm pitch FPC, FPC connector converter and the JXD320240A LCD (similar to the DMF682AN-EW Optrex LCD Module with D0-D4, FLM, LP, CP and M connectors) a few months back. See the pic below
![WhatsApp Image 2024-10-20 at 14 34 10](https://github.com/user-attachments/assets/c6cc29a5-c19f-4b97-90c7-430b3e039baf)

As I mentioned in my first ever post/reply, I am very very new to LCDs. I tried to connect the Fantom XR directly to the JXD320240A (without any PICO in between) and it didn't work. I was terribly confused and I gave up... until I saw your post/work on the Juno-G.

I have ordered a PICO and will try it out once it arrives. In the meantime, I may need to look up the datasheet with the JXD320240A LCD to see how it works.

Is there anything else I should look out for?

**Reply by bjaan** [reply]
Oh I see, I am not sure how to connect your specific screen to the Pico, and I would also think that you would be able to connect your Fantom XR directly to your screen based on the PIN informations on Aliexpress.

The description on Aliexpress appears to be very confusing as well, not sure what the comments are in Chinese, but it looks like the diagram is using different pins already, maybe a specific pin needs to be put High to enable the mode using the CP, LP, … mode

![image](https://github.com/user-attachments/assets/3ad752f2-f8e6-4a77-8518-8733ce9e5875)

This project was my first project working with graphical displays and the Pico, so I was very happy to get it working, and able to document it here as well. It was thanks to dpeddi‘s incredible work where I stumbled upon.

As your’s this is a different display, then I am using in this project, you should try to display something on it first using the Pico with a basic code example, it doesn’t have to be C++, and can be MicroPython. There are plenty of code libraries (APIs) for controlling displays available on-line, this project is also using one.

**Reply by maleksaid** [reply]
Thanks for the suggestions. I will definitely try some basic codes and libraries once the PICO arrives.

I totally agree with you that the description on AliExpress is rather confusing. I have absolutely no clue know how to reconcile between the **PIN Connection** tables to the **Block Diagram**. While waiting for the PICO to arrive, I will try once again to connect the Fantom XR to the LCD based on the PIN information.

In any case, I will update you what happens. Thanks for all your help thus far. Really appreciate it.
